### PR TITLE
feat: automated releases via github action

### PR DESCRIPTION
**Pull request by psychedelicious**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- [x] Yes
- [ ] No

## Description

This PR introduces a new, hands-off release process, which is handled almost entirely by GitHub actions. The only part that is not yet in an action is creating the GitHub release itself, but there are some notes suggesting a path to that (shouldn't be too hard).

This PR was motivated by some hiccups we've seen during releases:
- Missing frontend builds
- Complicated build process, run locally and subject to platform-specific quirks
- PyPI release action that could result in a different wheel being published to PyPI than what is released in the installer

See `docs/RELEASE.md` for documentation.

---

[feat: automated releases via github action](https://github.com/invoke-ai/InvokeAI/commit/48f364c5ac1a2e1f0adcc05f81c7b8792500da35)

- Restructure & update code check workflows
- Add release workflow to handle checks/tests, build and publish to PyPI
- Add docs/RELEASE.md explaining the workflow & process
- `create_installer.sh`: Update to work with the release workflow
- `create_installer.sh` & `tag_release.sh`: Fix the ANSI escape codes for macOS
- `tag_release.sh`: Add check for python binary name
- `tag_release.sh`: Print `git remote -v` output
- `tag_release.sh`: Fix error when deleting nonexistant tags

## QA Instructions, Screenshots, Recordings

This has been extensively tested on my fork and works great. Because the sensitive publish steps are protected in multiple ways, I think it's safe to merge and test on origin.

## Merge Plan

This PR needs review from @lstein. I'd also like to get reviews from @ebr and @brandonrising .

I'll defer to @lstein for whether or not to try this new process out for 3.5.0 or wait for the next release.

We can add up to 6 contributors to the list of approvers for the `pypi` and `testpypi` environments. Probably:
- @lstein 
- @Millu 
- @hipsterusername 
- @blessedcoolant 
- little ol' me?

## Added/updated tests?

- [ ] Yes
- [x] No : No code changes here

### CODE REVIEW

File: .github/actions/install-frontend-deps/action.yml
Comment (psychedelicious) [inline .github/actions/install-frontend-deps/action.yml:1]: This and the `install-python-deps` actions have been pulled into composite actions so they can be reused in workflows and we don't have to update multiple workflows if we need to change these steps.

File: .github/workflows/build-container.yml
Comment (psychedelicious) [inline .github/workflows/build-container.yml:14]: With the new release workflow triggering on tags, and use making two tags per release (e.g. `v3.5.0` and `v3-latest`), this was running twice. The narrower pattern ensures we only build the container once on release tags.

File: .github/workflows/check-frontend.yml
Comment (psychedelicious) [inline .github/workflows/check-frontend.yml:1]: `check-frontend`, `check-python` and `check-pytest` no longer trigger on commit/push/etc. They are only called by other workflows.

Separating the concepts of "run checks" and "should I run checks" means we can cleanly compose workflows.

For example, `on-change-check-frontend` triggers on commit/push/etc and does two things:
- Check if there are frontend files changed
- If so, run `check-frontend`

The `release` workflow also runs `check-frontend`, but doesn't care if files have changed or not - it *always* should run this workflow.

File: .github/workflows/on-change-check-frontend.yml
Comment (psychedelicious) [inline .github/workflows/on-change-check-frontend.yml:1]: As described in another comment, the `on-changed...` workflows are wrappers that conditionally run the actual checks. The `on-changed...` workflows are triggered by code changes.

File: .github/workflows/pypi-release.yml.old
Comment (psychedelicious) [inline .github/workflows/pypi-release.yml.old:1]: I've left this in as a reference but I suppose we have git history for that...

File: .github/workflows/release.yml
Comment (psychedelicious) [inline .github/workflows/release.yml:13]: The workflow runs on push of a tag matching `v*.*.*` _or_ when run manually. If it is run manually, you can skip the time-consuming code checks with this input.

File: .github/workflows/release.yml
Comment (psychedelicious) [inline .github/workflows/release.yml:1]: See `docs/RELEASE.md` for a detailed explanation of this workflow.

File: installer/create_installer.sh
Comment (psychedelicious) [inline installer/create_installer.sh:10]: My terminal emulator on macOS borked with `\e` - I had to change to `\033`, which works on all systems.

File: installer/create_installer.sh
Comment (psychedelicious) [inline installer/create_installer.sh:30]: `-v` doesn't work on some version of bash (like the one installed on my mac, apparently). This is a more generic way to do the same thing.

File: installer/create_installer.sh
Comment (psychedelicious) [inline installer/create_installer.sh:26]: I removed this because it was fiddly to make work with the workflows. The previous command ensures the directory is set correctly anyways.

File: installer/create_installer.sh
Comment (psychedelicious) [inline installer/create_installer.sh:80]: In CI, we want to emit both the wheel and source dist.

File: installer/create_installer.sh
Comment (psychedelicious) [inline installer/create_installer.sh:101]: Leave the wheel alone, so CI can get it.

File: installer/tag_release.sh
Comment (psychedelicious) [inline installer/tag_release.sh:62]: Added this sanity check - shows you what your current origin is.

File: installer/tag_release.sh
Comment (psychedelicious) [inline installer/tag_release.sh:70]: Two issues with `git push --delete origin $VERSION`:
- it fails if the tag doesn't exist
- if `$VERSION` happens to be a branch, it deletes it

The more obscure-looking `git push origin :refs/tags/$VERSION` has neither of these issues.

**Comment by hipsterusername** [comment]
Awesome - 👏

File: installer/tag_release.sh
Comment (ebr) [inline installer/tag_release.sh:13]: `command -v "$1" >/dev/null` is the usual way to do it

File: installer/tag_release.sh
Comment (ebr) [inline installer/tag_release.sh:28]: would the `python3` binary ever be missing if the system has Python 3 installed? Plus this doesn't work if the system also has Python 2. I feel like we should just be using `python3` directly, since Python can be finicky around being aliased.

**Review by ebr — APPROVED** [review]
Fantastic! :clap: This has been long overdue.
Left a few comments that are mostly nitpicks. There will probably be more kinks to work out in github actions once this is merged, but we won't know until the process gets some exercise.

File: .github/workflows/on-change-check-frontend.yml
Comment (ebr) [inline .github/workflows/on-change-check-frontend.yml:1]: Should the name of the workflow also convey when the workflow runs? If we enable it to run on dispatch, the name will be misleading. We should just name them after what they do: "Frontend checks", etc.

Comment (psychedelicious) [reply .github/workflows/on-change-check-frontend.yml:1]: I was trying to solve for the awkward interactions between how workflows get triggered without duplicating the workflow logic. 

For example, the pytest workflow should be run:
- on dispatch, always
- on call, always
- on code change, if the code change touches the app code

The changed files action complicates things by failing (not gracefully!) when used in a dispatched workflow because the last commit and target git ref are the same. It sees no changes and fails. 

First I tried using a series of "force" or "skip" inputs but this got really convoluted.

Then I learned about `github.event_name` (to only run on call/dispatch). But that's also messy, adding a lot of ifs. Maybe it's less messy than what I ended up with...

...which is splitting the workflow "business logic" from the "should I run this" logic. Unfortunately this is hard to cleanly represent and I ended up with two workflows for pytest:

1. a workflow that only runs on call or dispatch and always runs pytest
2. a workflow that runs said workflow on code changes

So there's a "Check: pytest" that only runs on call/dispatch and "On change: pytest" that runs on code changes.

(This all applies to the frontend and python code checks too)

What would be the conventional way to handle this situation? This is my first rodeo 😅

Comment (psychedelicious) [reply .github/workflows/on-change-check-frontend.yml:1]: Oh - it would be much easier if pytest could be a composite action - but it's a job with a matrix, so it's hard to compose...

Comment (ebr) [reply .github/workflows/on-change-check-frontend.yml:1]: thanks for explaining - the naming decision makes way more sense now.

> What would be the conventional way to handle this situation?

there isn't one :smiling_face_with_tear:  Every CI system likes its own way of doing things. With GHA, your experiences of `this got really convoluted.` > `that's also messy, adding a lot of ifs` > `this is hard to cleanly represent` is par for the course :)

I think there should be a way to further extract `pytest` jobs from workflows into actions the way you did with dependencies.... Need to process.

File: .github/workflows/on-change-check-frontend.yml
Comment (brandonrising) [inline .github/workflows/on-change-check-frontend.yml:33]: still pains me there isn't something built into actions for this

**Comment by lstein** [comment]
Sorry, I've been working on the model installer pretty intensively over the past few days. Will give this a review on Thursday.

File: .github/workflows/build-container.yml
Comment (lstein) [reply .github/workflows/build-container.yml:14]: Do we actually still need `v3-latest`? I introduced it during a period when we were still actively bug-fixing the 2.0 branch and we had reasons to tag the latest release on each branch, but that time is long past.

Comment (Millu) [reply .github/workflows/build-container.yml:14]: I think it's safe to remove

**Review by lstein — APPROVED** [review]
Looks like this has been picked over very carefully already. I scanned through and couldn't find anything that looked suspicious, so I'm giving it my go ahead. Impressive work!

File: installer/tag_release.sh
Comment (psychedelicious) [reply installer/tag_release.sh]: Yeah, that's much simpler, thanks! Changed.

File: .github/workflows/on-change-check-frontend.yml
Comment (psychedelicious) [reply .github/workflows/on-change-check-frontend.yml:6]: Let's keep this in mind for future enhancements

File: .github/workflows/build-container.yml
Comment (psychedelicious) [reply .github/workflows/build-container.yml:14]: Removed

**Comment by psychedelicious** [comment]
@lstein @hipsterusername I think this will need to be merged with the big red button bc it changes workflows.

And then the repo settings will need to be updated so that PRs require these three workflows to succeed:
- `On change: run check-frontend`
- `On change: run check-python`
- `On change: run pytest`

Otherwise this is good to go, and we can test it with a dev/alpha release (that we don't publish)

**Comment by hipsterusername** [comment]
this is great. going to say we hold off on this until after 3.6 is out

**Comment by hipsterusername** [comment]
Want to merge this but going to wait until we have a few folks ready to coordinate the other changes needed for this to succeed

**Comment by psychedelicious** [comment]
Yeah we should coordinate. I'd like to find a half hour to merge this and do a dev release to pypi on a call with @lstein. Just in case we need to do pypi surgery. 

Just had a thought - does the updater only offer to update to RC versions? If we release a dev/nightly version to pypi will users be prompted to update?

Oooh. Nightlies :) that sounds cool.

**Comment by Millu** [comment]
> Just had a thought - does the updater only offer to update to RC versions? If we release a dev/nightly version to pypi will users be prompted to update?

The updater right now only looks for RCs and the latest release, but could be easily updated for nightlies

**Comment by psychedelicious** [comment]
The updater logic currently only differentiates between RC and non-RC.

This means alpha, dev, and post releases are all considered non-RC releases, and if we do an alpha or dev release, a users running the updater at the right moment will see it as a stable release and download it.

I'll post a function to split the versions up more granularly in the #5507, but this may be tricky to now that users have this in their updater script.

**Comment by psychedelicious** [comment]
I've updated this branch. 

Post-merge testing: 
Testing:
- Manually dispatch the Release workflow. Confirm it builds a working wheel.
- Create a PR that changes frontend files and should fail a lint check.
- Create a PR that changes frontend files and should pass a lint check.
- Create a PR that changes python files and should fail a lint check.
  - It should also trigger the pytest workflow
- Create a PR that changes python files and should pass a lint check.
  - It should also trigger the pytest workflow
- Create a branch that tags a new release. Use the make tag-release command to tag the release accordingly. This should kick off the Release workflow, running all checks. It will pause before anything is published.
  - If all goes well, we can authorize only the testpypi publish step to publish the release to testpypi.
  - Test the testpypi release (instructions in RELEASE.md)

If all of these steps work correctly, theoretically we should be ready for a real release when the time comes.

**Review by hipsterusername — APPROVED** [review]

**Comment by psychedelicious** [comment]
I've updated this PR to work with the recent changes to the install process.

**Comment by psychedelicious** [comment]
Superseded by https://github.com/invoke-ai/InvokeAI/pull/5839

**Pull request closed without merging**
